### PR TITLE
[CIT-500] thin edge json extension to handle software management messages

### PR DESCRIPTION
**Pull request by didier-wenzek**
## Proposed changes

* Helper methods to be used by the clould mappers and the sm_agent.
* The purpose is to encapsulate the actual message format.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: sm/json_sm/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                            "name": "mongodb",
+                            "version": "4.4.6",
+                            "action": "remove"
+                        }
+                    ]
+                }
+            ]
+        }"#;
+        let actual_json = request.to_json().expect("Failed to serialize");
+        assert_eq!(actual_json, remove_whitespace(expected_json));
```
Comment (makr11st) [inline sm/json_sm/src/lib.rs:174]: I believe you don't need this function `remove_whitespace`, you could use `split_whitespace`, something like:
```suggestion
        assert_eq!(actual_json, expected_json.split_whitespace().collect::<String>());
```

File: sm/json_sm/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "json_sm"
+version = "0.1.0"
```
Comment (makr11st) [inline sm/json_sm/Cargo.toml:3]: I suppose the version should  be align with the rest of tedge.
```suggestion
version = "0.2.3"
```

File: sm/json_sm/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    },
+}
+
+impl From<serde_json::Error> for SoftwareError {
+    fn from(err: serde_json::Error) -> Self {
+        SoftwareError::ParseError {
+            reason: format!("{}", err),
+        }
+    }
+}
```
Comment (makr11st) [inline sm/json_sm/src/error.rs:63]: Isn't this duplication of lines 15-16?

File: sm/json_sm/src/messages.rs [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+        SoftwareRequestResponse {
+            id,
+            status,
+            current_software_list: vec![],
+            reason: None,
+            failures: vec![],
+        }
+    }
+
+    pub fn finalize_response(&mut self, software_list: Vec<SoftwareRequestResponseSoftwareList>) {
```
Comment (makr11st) [inline sm/json_sm/src/messages.rs:189]: I suppose it is a leftover from my code but I suppose `software_list` should be a ref to Vec and not Vec, right?
Or maybe even better a slice?
```suggestion
    pub fn finalize_response(&mut self, software_list: &[SoftwareRequestResponseSoftwareList]) {
```

File: sm/json_sm/Cargo.toml (hunk 2 above)
Comment (didier-wenzek) [reply sm/json_sm/Cargo.toml:3]: Fixed

**Comment by rina23q** [comment]
Where is the struct corresponding to executing? This message comes from agent.

```
{
    "id": 123,
    "status": "EXECUTING"
}
```

**Review by rina23q — COMMENTED** [review]
Where is the struct corresponded to executing? This message comes from agent.

```
{
    "id": 123,
    "status": "executing"
}
```

**Comment by didier-wenzek** [comment]
> Where is the struct corresponded to executing?

This is a `SoftwareUpdateResponse`.
* This struct is created in the `executing` status.
* Sent by the agent to the mapper as `{ "id": ... , "status": "executing"}`
* Then populated by the agent once the install is done with the new software list and the errors if any

**Review by rina23q — APPROVED** [review]
We will polish while implementing agent and mapper ;)

**Pull request merged**